### PR TITLE
Add code coverage and coveralls

### PR DESCRIPTION
**Pull request by gangliao**
@wangkuiyi @reyoung @hedaoyuan  Requirement from #255.
This PR is completely implemented in CMake to support code coverage which mainly include three components.

1.  filter source code for code coverage, such as,  *.cc, *.cpp, etc.
2.  `make test` and generate gcov data
3. convert *.gcov to json and upload it to coveralls.
 https://coveralls.io/github/baidu/Paddle 
[Coverage Status](https://coveralls.io/builds/8602716)

Currently, Paddle achieves 63% coverage, way to go !!!

If you want to do code coverage in your local machine, you can simply issue the following commands:
```bash
cmake -DON_COVERALLS=ON -DCOVERALLS_UPLOAD=OFF ..
make
make coveralls
```

### CODE REVIEW

File: paddle/scripts/travis/build_and_test.sh [hunk 1]
```diff
@@ -6,17 +6,19 @@ if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then
...
 
 
-cmake .. -DCMAKE_BUILD_TYPE=Debug -DWITH_GPU=OFF -DWITH_DOC=OFF -DWITH_TESTING=ON -DON_TRAVIS=ON ${CMAKE_EXTRA}
+cmake .. -DCMAKE_BUILD_TYPE=Debug -DWITH_GPU=OFF -DWITH_DOC=OFF -DWITH_TESTING=ON -DON_TRAVIS=ON -DON_COVERALLS=ON ${CMAKE_EXTRA}
 
 NPROC=1
 if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
   NRPOC=`nproc`
+  make -j $NPROC
+  make coveralls
```
Comment (reyoung) [inline paddle/scripts/travis/build_and_test.sh]: 1、这个converalls会不会在Fail的时候，打出log来？
2、这个converalls是不是多线程的？

Comment (gangliao) [inline paddle/scripts/travis/build_and_test.sh]: 恩 确实得改成多线程的

看了一下log，exclude还得过滤一些文件。。。
https://travis-ci.org/baidu/Paddle/jobs/172015221#L5605

**Comment by gangliao** [comment]
According to the log https://s3.amazonaws.com/archive.travis-ci.org/jobs/172015221/log.txt,
some bugs and drawbacks will be fixed tomorrow.

File: paddle/scripts/travis/build_and_test.sh (hunk 1 above)
Comment (gangliao) [inline paddle/scripts/travis/build_and_test.sh]: 多线程已添加 @reyoung

File: cmake/coveralls.cmake [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        ".test_env/"
+      )
+
+    if(WITH_GPU)
+        file(GLOB_RECURSE PADDLE_SOURCES RELATIVE "${PROJECT_SOURCE_DIR}" "*.cpp" "*.cc" ".c" "*.cu")
+    else()
+        file(GLOB_RECURSE PADDLE_SOURCES RELATIVE "${PROJECT_SOURCE_DIR}" "*.cpp" "*.cc" "*.c")
+    endif()
+
+    # exclude trivial files in PADDLE_SOURCES
```
Comment (gangliao) [inline cmake/coveralls.cmake]: 过滤

**Comment by reyoung** [comment]
Does `COVERALLS` have account or key to login? If not, is that means anyone could post a fake `CONVERALLS` stats?

**Comment by gangliao** [comment]
@reyoung  `Coveralls` needs to associate with github indeed and it also need our organization's authorization like `travis`.  All these thing already done three days ago.

File: cmake/coverallsGcovJsons.cmake [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+# 3. Run this script specifying which source files the coverage should be performed on.
+#
+# This script will then use gcov to generate .gcov files in the directory specified
+# via the COV_PATH var. This should probably be the same as your cmake build dir.
+#
+# It then parses the .gcov files to convert them into the Coveralls JSON format:
+# https://coveralls.io/docs/api
+#
+
+CMAKE_MINIMUM_REQUIRED(VERSION 2.8)
```
Comment (reyoung) [inline cmake/coverallsGcovJsons.cmake]: This line may be incompatible with #212.
Cmake on `centos 7` maybe too old. @QiJune please check  centos is cool for this.

File: cmake/coverallsGcovJsons.cmake [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+	message(FATAL_ERROR "gcov not found! Aborting...")
+endif()
+
+find_package(Git)
+
+# TODO: Add these git things to the coveralls json.
+if (GIT_FOUND)
+	# Branch.
+	execute_process(
+		COMMAND ${GIT_EXECUTABLE} rev-parse --abbrev-ref HEAD
```
Comment (reyoung) [inline cmake/coverallsGcovJsons.cmake]: Git Branch is an environment variable in travis-ci.  See [here](https://github.com/baidu/Paddle/blob/develop/paddle/scripts/travis/docs.sh#L25)

File: cmake/coverallsGcovJsons.cmake [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+	git_log_format(ce GIT_COMMITTER_EMAIL)
+	git_log_format(B GIT_COMMIT_MESSAGE)
+
+	message("Git exe: ${GIT_EXECUTABLE}")
+	message("Git branch: ${GIT_BRANCH}")
+	message("Git author: ${GIT_AUTHOR_NAME}")
+	message("Git e-mail: ${GIT_AUTHOR_EMAIL}")
+	message("Git commiter name: ${GIT_COMMITTER_NAME}")
+	message("Git commiter e-mail: ${GIT_COMMITTER_EMAIL}")
+	message("Git commit message: ${GIT_COMMIT_MESSAGE}")
```
Comment (reyoung) [inline cmake/coverallsGcovJsons.cmake]: All these things could be get by travis-ci.

Comment (reyoung) [inline cmake/coverallsGcovJsons.cmake]: But it is still cool now. No need to change here.

**Pull request merged**